src/main/java/org/skriptlang/skript/bukkit/recipes/RecipeWrapper.java

```diff
@@ -0,0 +1,385 @@
+package org.skriptlang.skript.bukkit.recipes;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.*;
+import org.bukkit.inventory.recipe.CookingBookCategory;
+import org.bukkit.inventory.recipe.CraftingBookCategory;
+import org.skriptlang.skript.bukkit.recipes.RecipeUtils.RecipeType;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Wrapper class allowing the creation of a Recipe without being on initialization.
+ */
+public abstract class RecipeWrapper implements Recipe {
+
+	private final RecipeType recipeType;
+	private final NamespacedKey key;
+	private ItemStack result = null;
+	private List<String> errors = new ArrayList<>();
+
+	public RecipeWrapper(NamespacedKey key, RecipeType recipeType) {
+		this.key = key;
+		this.recipeType = recipeType;
+	}
+
+	public void setResult(ItemStack result) {
+		this.result = result;
+	}
+
+	@Override
+	public ItemStack getResult() {
+		return result;
+	}
+
+	public NamespacedKey getKey() {
+		return this.key;
+	}
+
+	public RecipeType getRecipeType() {
+		return recipeType;
+	}
+
+	public void addError(String error) {
+		errors.add(error);
+	}
+
+	public List<String> getErrors() {
+		return errors;
+	}
+
+	public abstract Recipe create();
+
+    public static class CraftingRecipeWrapper extends RecipeWrapper {
+		private RecipeChoice[] ingredients = new RecipeChoice[9];
+		private String group;
+		private CraftingBookCategory category;
+
+		public CraftingRecipeWrapper(NamespacedKey key, RecipeType recipeType) {
+			super(key, recipeType);
+		}
+
+		@Override
+		public Recipe create() {
+			return null;
+		}
+
+		public void setIngredients(int placement, RecipeChoice item) {
+			ingredients[placement] = item;
+		}
+
+		public void setGroup(String group) {
+			this.group = group;
+		}
+
+		public void setCategory(CraftingBookCategory category) {
+			this.category = category;
+		}
+
+		public RecipeChoice[] getIngredients() {
+			return ingredients;
+		}
+
+		public String getGroup() {
+			return this.group;
+		}
+
+		public CraftingBookCategory getCategory() {
+			return this.category;
+		}
+
+		public static class ShapedRecipeWrapper extends CraftingRecipeWrapper {
+
+			public ShapedRecipeWrapper(NamespacedKey key, RecipeType recipeType) {
+				super(key, recipeType);
+			}
+
+			@Override
+			public ShapedRecipe create() {
+				if (getResult() == null) {
+					addError("You must provide a result item when registering a recipe");
+					return null;
+				}
+				RecipeChoice[] ingredients = getIngredients();
+				if (ingredients.length == 0 || Arrays.stream(ingredients).filter(Objects::nonNull).toArray().length < 2) {
+					addError("You must have at least 1 ingredient when registering a '" + getRecipeType() + "' recipe.");
```

Comment:
this checks for 
```suggestion
				if (ingredients.length == 0 || Arrays.stream(ingredients).filter(Objects::nonNull).count() < 2) {
					addError("You must have at least 1 ingredient when registering a '" + getRecipeType() + "' recipe.");
```
is it 1 ingredient or 2 required
because the code says 2 but the text says 1